Bluetooth: controller: Fix ISO broadcaster pre-transmission groups > 1

mtpr-ot opened the pull request:
For pre-transmission groups > 1, the broadcaster link layer would fetch incorrect payloads from the TX node queue.

Update payload index calculation to reference correct TX payload. Fix PTO FIXMEs and range checks.

bsim bap_broadcast_audio_source/sink tested with PTO=2 PTO Group Count = 2:
[bsim_bap_broadcast_source_sink_pto_2_ptogc_2.zip](https://github.com/user-attachments/files/18363537/bsim_bap_broadcast_source_sink_pto_2_ptogc_2.zip)

mtpr-ot committed:
Bluetooth: controller: Fix ISO broadcaster pre-transmission groups > 1

For pre-transmission groups > 1, the broadcaster link layer would fetch
incorrect payloads from the TX node queue.

Update payload index calculation to reference correct TX payload. Fix
PTO FIXMEs and range checks.

Signed-off-by: Morten Priess <mtpr@oticon.com>

cvinayak commented:
@mtpr-ot This PR was never put to ready for review, intentional? Seems this is valid change request, if there is no objection, I will put this for review and follow up (provided can do some testing if possible)

Tronil commented:
Hi @cvinayak - Morten recently left Oticon, so I'm not sure if he is still active with this account.

In any case, we are using this change downstream, so I think it is just an oversight that this PR never left draft.

cvinayak commented:
> Hi @cvinayak - Morten recently left Oticon, so I'm not sure if he is still active with this account.
> 
> In any case, we are using this change downstream, so I think it is just an oversight that this PR never left draft.

Ok, I will follow up on the failing bsim CI test, suspect an overlapping/drifting ACL role leading to missing ISO SDU reception...

cvinayak committed:
Bluetooth: Controller: Add back the use of pre-transmissions

Add back the implementation in the Controller that tries to
enable pre-transmissions to improve time diversity to aid a
remote ISO Sync Receiver role device.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -627,9 +627,24 @@ static void isr_tx_common(void *param,
+			ptx_group_mult = (ptx_group_idx + 1) * lll->pto;
+			payload_index = ptx_payload_idx + ptx_group_mult * lll->bn;
+		} else {
+			payload_index  = lll->bn_curr - 1U;
```
```suggestion
			payload_index = lll->bn_curr - 1U;
```

Any risk of underflow here?

cvinayak replied:
No, bn_curr range of 1 to 7.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -627,9 +627,24 @@ static void isr_tx_common(void *param,
+			payload_index  = lll->bn_curr - 1U;
+		}
+
+		payload_count = lll->payload_count - lll->bn + payload_index;
```
Similarly can `lll->bn` be larger than `lll->payload_count` here?

cvinayak replied:
No, `lll->payload_count` was pre-incremented by `lll->bn`  here: https://github.com/zephyrproject-rtos/zephyr/blob/42faa9438f8b0ddb827718b3d20699f9bf88e0c0/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c#L206-L207

Thalley replied:
If `lll->latency_prepare` then that comment doesn't hold up. Can `lll->latency_prepare == 0`?

cvinayak replied:
`lll->latency_prepare` is greater than `1` when the `lll->payload_count` is incremented.

https://github.com/zephyrproject-rtos/zephyr/blob/966122045b9f555373db6b6eeed53cc0a84d76ea/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c#L117-L123

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -1155,17 +1162,39 @@ static uint8_t ptc_calc(const struct lll_adv_iso *lll, uint32_t event_spacing,
 
-		/* FIXME: Here we restrict to a maximum of BN Pre-Transmission
-		 * subevents per BIS
+		/* Required NSE */
+		nse = lll->bn * lll->irc;
```
Ditto here; risk of overflow when multiplying 2 uint8_ts?

cvinayak replied:
No, `bn` is 3 bits and `irc` is 4 bits.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -627,9 +627,24 @@ static void isr_tx_common(void *param,
+			ptx_group_idx = ptx_idx / lll->bn;
+			ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn;
+			ptx_group_mult = (ptx_group_idx + 1) * lll->pto;
+			payload_index = ptx_payload_idx + ptx_group_mult * lll->bn;
```
Fixed.

cvinayak committed:
Bluetooth: controller: Fixup Broadcast ISO pre-transmission groups > 1

Fixup payload_index to be uint16_t to avoid index overflow.

Do not remember why ptc is 4 bits, where as it must be 5 bit
value similar to nse; added an assertion check until it is
fixed.

Fix ISO Broadcaster and ISO Sync Receiver for PTO > 1 and
use of Pre-Transmission Group Counts.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -625,11 +625,27 @@ static void isr_tx_common(void *param,
-				(lll->ptc_curr * lll->pto);
-		payload_count = lll->payload_count + payload_index - lll->bn;
+		if (lll->ptc_curr) {
+			uint8_t ptx_idx = lll->ptc_curr - 1; /* max. nse 5 bits */
+			uint8_t ptx_payload_idx;
+			uint16_t ptx_group_mult;
+			uint8_t ptx_group_idx;
+
+			/* Calculate group index and multiplier for deriving
+			 * pre-transmission payload index.
+			 */
+			ptx_group_idx = ptx_idx / lll->bn; /* 5 bits */
+			ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn; /* 8 bits */
+			ptx_group_mult = (ptx_group_idx + 1) * lll->pto; /* 9 bits */
+			payload_index = ptx_payload_idx + ptx_group_mult * lll->bn; /* 13 bits */
```
The comments are somewhat helpful. Currently they are too vague (I know from context that they refer to the maximum size of the result, but that's not clear for someone else reading this later). 

As you may be aware, this is a general concern when I'm reviewing controller PRs, and in some cases there are potential issues caught and found (e.g. you've modified the type of `ptx_group_mult` and `payload_index` based on my review). 

The comments are OK assuming that the size of these values never change. A more future proof approach would be to have asserts ensuring that we don't have silent and unintentional under- or overflows.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -625,11 +625,27 @@ static void isr_tx_common(void *param,
-				(lll->ptc_curr * lll->pto);
-		payload_count = lll->payload_count + payload_index - lll->bn;
+		if (lll->ptc_curr) {
+			uint8_t ptx_idx = lll->ptc_curr - 1; /* max. nse 5 bits */
```
Since `ptc_curr` is `uint32_t ptc_curr:4;`, then shouldn't this be 
```suggestion
			uint8_t ptx_idx = lll->ptc_curr - 1; /* max. nse 4 bits */
```

Which then also affect the bits of several of the values below?

Tronil replied:
> Now I do not remember why `ptc` is 4 bits! ... it can be as much as (NSE - 1); 1 transmission followed by (NSE - 1) pre-transmissions.
> 
> @Tronil and @wopu-ot , are you testing the boundary conditions internally in your tests for this PR changes?

We do not have extensive test coverage on the broadcaster side, no

cvinayak replied:
@Tronil while testing this PR for PTO=2 and Group Count = 2, I discovered a bug in ull_sync_iso.c that subsequent BIS subevent reception failed; `ptc` was not assigned the correct value. This bug will not show up if only one BIS is synchronized to.

Thalley replied:
Shouldn't the comment be updated?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -625,11 +625,27 @@ static void isr_tx_common(void *param,
+			 * pre-transmission payload index.
+			 */
+			ptx_group_idx = ptx_idx / lll->bn; /* 5 bits */
+			ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn; /* 8 bits */
```
> hmmm... isnt  ptx_payload_idx = ptx_idx - ((ptx_idx / lll->bn) * lll->bn; always zero! ?

I though so too at first, but due to integer division it isn't. 

e.g. pts_index = 5 and lll->bn = 3 then
```
ptx_group_idx = ptx_idx / lll->bn; /* will be 1 */ 
ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn; /* will be 2 */
```

If `ptx_group_idx = ptx_idx / lll->bn;` was not integer divided and resulted in 5/3, then `ptx_payload_idx` would have become 0. 

I expected that was intentional, but I'm not sure

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -625,10 +625,26 @@ static void isr_tx_common(void *param,
+			ptx_group_idx = ptx_idx / lll->bn; /* 5 bits */
+			ptx_payload_idx = ptx_idx - ptx_group_idx * lll->bn; /* 8 bits */
+			ptx_group_mult = (ptx_group_idx + 1U) * lll->pto; /* 9 bits */
+			payload_index = ptx_payload_idx + ptx_group_mult * lll->bn; /* 13 bits */
+			LL_ASSERT(payload_index <= UINT8_MAX);
```
Something looks a bit here now; you have a comment stating `13 bits` but you also assert that it's less <= 8 bits

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -1336,6 +1335,31 @@ static void isr_done(void *param)
+		payload_index  = lll->bn_curr - 1U; /* 3 bits */
+	}
+
+	return payload_index & 0xFF;
```
If you want to ensure that you return the correct type, you should do this:

```suggestion
	return (uint8_t)payload_index;
```

`0xFF` is an `int` so you are effectively implicitly casting the type to an `int` here, and you already have an assert for the correct value range above

cvinayak replied:
Updated code. ISO Sync Receiver has no data type limitation that is applicable for ISO Broadcaster wrt buffer indices (runtime check for overflows are present).

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -201,12 +212,11 @@ static uint8_t big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bi
 
-			/* FIXME: PTO is currently limited to BN */
-			if (!IN_RANGE(pto, 0x00, bn /*0x0F*/)) {
+			if (!IN_RANGE(pto, 0x00, 0x0F)) {
```
Coverity will complain about `IN_RANGE` on unsigned integers when the first value is 0x00. 

```suggestion
			if (pto > 0x0F)) {
```

cvinayak replied:
Fixed.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -1155,19 +1173,51 @@ static uint8_t ptc_calc(const struct lll_adv_iso *lll, uint32_t event_spacing,
-		/* FIXME: Here we restrict to a maximum of BN Pre-Transmission
-		 * subevents per BIS
+		/* Required NSE */
+		nse = lll->bn * lll->irc; /* 3 bits + 4 bits */
```
```suggestion
		nse = lll->bn * lll->irc; /* 3 bits * 4 bits */
```

cvinayak replied:
Fixed.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -1155,19 +1173,51 @@ static uint8_t ptc_calc(const struct lll_adv_iso *lll, uint32_t event_spacing,
 		 */
-		ptc = MIN(ptc, lll->bn);
+		LL_ASSERT(ptc <= BIT_MASK(4));
 
-		return ptc;
+		return ptc & BIT_MASK(5);
```
Is the `& BIT_MASK(5)` valuable when you already have the assert?

cvinayak replied:
agree, fixed.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -483,13 +483,20 @@ void ull_sync_iso_setup(struct ll_sync_iso_set *sync_iso,
-		lll->ptc = lll->bn;
+		uint8_t nse;
+
+		nse = lll->irc * lll->bn; /* 4 bits + 3 bits */
```
```suggestion
		nse = lll->irc * lll->bn; /* 4 bits * 3 bits */
```

cvinayak replied:
Fixed.

cvinayak reviewed (approved):
I am a co-author is this PR, but do approve the initial commits from the original author.
